src/ripple/app/misc/Validations.cpp

```diff
@@ -104,39 +104,67 @@ class ValidationsImp : public Validations
         }
 
         auto hash = val->getLedgerHash ();
-        auto node = val->getNodeID ();
 
-        if (val->isTrusted () && isCurrent)
+        if (isCurrent &&
+            (val->isTrusted () || app_.validators ().listed (signer)))
         {
             ScopedLockType sl (mLock);
 
-            if (!findCreateSet (hash)->insert (std::make_pair (node, val)).second)
+            auto const masterKey = app_.manifestCache ().getMasterKey (signer);
```

Comment:
I started by observing that we use this idiom in a number of places: call `getMasterKey` with some input, then depending on the result itself, we use either the result or the original input. So I was going to suggest, why not simply have `getMasterKey` return `signer` and avoid this conditional?

But then it occurred to me, what if the `signer` here is a just-revoked key? Notice that we call `listed` without any locks.

Can you explain the logic here and why there isn't a race?